Solved Data-Races

hernanponcedeleon opened the pull request:
Unfortunately #1157 did not solve all the data races as I expected. Due to a bug in Dartagnan which resulted in an incomplete analysis I did not detected this before. This PR fixes the remaining problems. Besides manual inspection, it would be nice if we could run any of the other tools that can detect data races (@sim642 or @MartinSpiessl could you please help with this?) to be sure that the claims in this PR about the expected results are correct.

Also, the changes introduced in #1157 modified the original behaviour of two benchmarks (`read_write_lock-2.c` and `race-3_2-container_of-global.c`). The goal of the changes was to wrap instructions with atomic_begin/end. However some of the wrapped instructions corresponded to two memory accesses `assert(y == x);` (read the value of both y and x) and `data->shared.b = data->shared.b + 1;` (first read and then write the value of shared.b). This PR restores the original behaviour by explicitly making each load/store and surrounded them individually by the atomic block to prevent the data race.

There are still some benchmarks with data races in `pthread-ext`, but as mentioned in #1176, they are duplicated and I think we should get rid of them.

I think the modified benchmarks of this PR are a clear example of why we should eventually moved to C11 atomics. We are practically wrapping every single memory access with an atomic_begin/end to prevent data races and the files become unreadable.

dbeyer commented:
@MartinSpiessl Would you be able to review this until the weekend?

dbeyer commented:
Or @sim642 could you help may be with reviewing this?

sim642 reviewed (commented):
I'm not too familiar with the fine-grained use of atomics, so I can't really argue where it needs to be added and where not. Also Goblint doesn't do such precise analysis (and currently also doesn't implement `__VERIFIER_atomic_*()`), so no insights from that.

Nevertheless, I did manually look through the modified files and left comments about two kinds of things that seemed suspicious:
1. Compared to the racing `-b` versions of the tasks, I see that some busy loops were rewritten to first atomically read a variable into a local but keep the loop as a whole non-atomic. I think in many places this produces a notable semantic difference because where previously the loop could at some point terminate, it now doesn't if it enters the loop at all because the local variable isn't atomically copied again to evaluate the loop condition with updated value.
2. Some of the benchmarks have two symmetric threads but there are differences regarding atomic blocks between them. Not sure if they should still be fully symmetric or they are actually already correct with a single atomic block missing in one thread.

P.S. Some of my comments are not exactly on the right lines because GitHub only let me put them near modified lines but these semantical differences were already introduced before.

dbeyer commented:
@hernanponcedeleon Could you please have a look?

kfriedberger commented on c/ldv-races/race-3_2-container_of-global.i:
```diff
@@ -18,49 +18,48 @@ typedef __int32_t __int_least32_t;
 typedef char *__caddr_t;
-__extension__ typedef int __intptr_t;
-__extension__ typedef unsigned int __socklen_t;
+typedef long int __intptr_t;
```
The change of `intptr_t` shows a change in the machine model. Please use 32bit instead of 64bit, as required for this category.

kfriedberger commented on c/ldv-races/race-3_2-container_of-global.i:
```diff
@@ -294,14 +292,14 @@ typedef unsigned int pthread_key_t;
 union pthread_attr_t
 {
-  char __size[36];
+  char __size[56];
```
Is this change related to machine model or other changes in C-Pthread header files?

kfriedberger commented on c/ldv-races/race-3_2-container_of-global.i:
```diff
@@ -294,14 +292,14 @@ typedef unsigned int pthread_key_t;
 {
   struct __pthread_mutex_s __data;
-  char __size[24];
+  char __size[40];
```
same here.

kfriedberger commented on c/ldv-races/race-3_2-container_of-global.i:
```diff
@@ -313,7 +311,7 @@ typedef union
 {
   struct __pthread_rwlock_arch_t __data;
-  char __size[32];
+  char __size[56];
```
and here.

kfriedberger commented on c/ldv-races/race-3_2-container_of-global.i:
```diff
@@ -324,15 +322,15 @@ typedef union
   int __align;
 } pthread_barrierattr_t;
-typedef int __jmp_buf[6];
+typedef long int __jmp_buf[8];
```
yep, looks like wrong machine model.

kfriedberger commented on c/pthread-atomic/read_write_lock-2.i:
```diff
@@ -36,49 +36,48 @@ typedef __int32_t __int_least32_t;
 typedef char *__caddr_t;
-__extension__ typedef int __intptr_t;
-__extension__ typedef unsigned int __socklen_t;
+typedef long int __intptr_t;
```
Machinemodel: 32 vs 64bit?

kfriedberger commented on c/pthread-ext/40_barrier_vf.c:
```diff
@@ -43,7 +43,7 @@ __VERIFIER_atomic_begin(); \
   (c = 1); \
-  __VERIFIER_atomic_begin(); \
+  __VERIFIER_atomic_end(); \
   } //BP must be post-processed manually by changing "b*_COND := 1" to "b*_COND$ := 1"
```
interesting: that is a bugfix and not related to data-race.

kfriedberger commented on c/pthread-ext/40_barrier_vf.i:
```diff
@@ -718,7 +718,7 @@ void Barrier2() {
   if (count == 3) {
-    { __VERIFIER_atomic_begin(); (COND = 1); __VERIFIER_atomic_begin(); };
+    { __VERIFIER_atomic_begin(); (COND = 1); __VERIFIER_atomic_end(); };
     count = 0; }
```
interesting: that is a bugfix and not related to data-race.

kfriedberger reviewed (changes requested):
This PR inserts atomic locks for several read and write accesses to global variables.
Additionally, some invalid atomic locks/unlocks are fixed.

The preprocessing might have been done with the wrong machine model. That should not be an problem for the benchmarks, they are mostly machine independent. However, it should be fixed for consistency with the yml-definition (`data_model: ILP32`).

hernanponcedeleon commented:
1. You were completely right @sim642, good catch! Since Dartagnan cannot prove correctness for such programs (because of `while(1)`), I did not notice this error. Besides the cases you pointed out, I fixed a couple more (e.g. in `Peterson.c`).
2. I went for the conservative solution and added `atomic_begin/end` to every single memory access. Syntactically, the new benchmarks look awful, but we will get rid of this once we use C11 atomics (hoping to start working on this when the benchmarks for 2021 are frozen).

@kfriedberger you are right, I forgot to add the `-m32` last time I pre-processed the files.

sim642 reviewed (approved):
@hernanponcedeleon Thanks for fixing the loops in other similar files as well! I only looked at the ones which this PR previously modified.

dbeyer commented:
@kfriedberger Satisfied?

kfriedberger reviewed (approved):
The last commit applied preprocessing with the correct machine model.
It also includes changes from the previous commit, where preprocessed files were not yet updated. 

PR seems to be acceptable now.
